components/p3a/message_manager.cc

```diff
@@ -0,0 +1,344 @@
+#include "base/strings/string_number_conversions.h"
+#include "brave/components/p3a/features.h"
+#include "brave/components/p3a/metric_names.h"
+#include "brave/components/p3a/p3a_config.h"
+#include "brave/components/p3a/pref_names.h"
+#include "brave/components/p3a/rotation_scheduler.h"
+#include "brave/components/p3a/scheduler.h"
+#include "brave/components/p3a/star_helper.h"
+#include "brave/components/p3a/star_log_store.h"
+#include "brave/components/p3a/uploader.h"
+#include "components/metrics/log_store.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+
+namespace p3a {
+
+namespace {
+
+const size_t kMaxEpochsToRetain = 4;
+constexpr base::TimeDelta kPostRotationUploadDelay = base::Seconds(30);
+
+bool IsSTAREnabled() {
+  return base::FeatureList::IsEnabled(features::kSTAR);
+}
+
+}  // namespace
+
+MessageManager::MessageManager(PrefService* local_state,
+                               const P3AConfig* config,
+                               Delegate* delegate,
+                               std::string channel,
+                               std::string week_of_install)
+    : local_state_(local_state), config_(config), delegate_(delegate) {
+  message_meta_.Init(local_state, channel, week_of_install);
+
+  // Init log stores.
+  for (MetricLogType log_type : kAllMetricLogTypes) {
+    json_log_stores_[log_type] =
+        std::make_unique<MetricLogStore>(this, local_state_, false, log_type);
+    json_log_stores_[log_type]->LoadPersistedUnsentLogs();
+  }
+  star_prep_log_store_ = std::make_unique<MetricLogStore>(
+      this, local_state_, true, MetricLogType::kTypical);
+  star_prep_log_store_->LoadPersistedUnsentLogs();
+  star_send_log_store_ =
+      std::make_unique<StarLogStore>(local_state_, kMaxEpochsToRetain);
+}
+
+MessageManager::~MessageManager() = default;
+
+void MessageManager::RegisterPrefs(PrefRegistrySimple* registry) {
+  MetricLogStore::RegisterPrefs(registry);
+  StarLogStore::RegisterPrefs(registry);
+  StarHelper::RegisterPrefs(registry);
+  RotationScheduler::RegisterPrefs(registry);
+}
+
+void MessageManager::Init(
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory) {
+  // Init other components.
+  uploader_ = std::make_unique<Uploader>(
+      url_loader_factory,
+      base::BindRepeating(&MessageManager::OnLogUploadComplete,
+                          base::Unretained(this)),
+      config_.get());
+
+  for (MetricLogType log_type : kAllMetricLogTypes) {
+    json_upload_schedulers_[log_type] = std::make_unique<Scheduler>(
+        base::BindRepeating(&MessageManager::StartScheduledUpload,
+                            base::Unretained(this), false, log_type),
+        config_->randomize_upload_interval, config_->average_upload_interval);
+    json_upload_schedulers_[log_type]->Start();
+  }
+  star_prep_scheduler_ = std::make_unique<Scheduler>(
```

Comment:
same as below, why are we initializing this stuff (and starting the scheduler) without first checking to see if star is enabled?